### PR TITLE
SnapshotTransaction stream refactor proof-of-concept

### PR DESCRIPTION
**Pull request by wi11dey**
Could probably also separate out all the postfiltering logic into its own `PostFilter` class

Birthday present for @rhuffy 

## General
**Before this PR**:

**After this PR**:

==COMMIT_MSG==
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 1]
```diff
@@ -1681,125 +1627,126 @@ private static boolean isSweepSentinel(Value value) {
...
-                    MoreExecutors.directExecutor());
+                                tableRef, getStartTimestamp(), nextRawResults.size() == keysToReloadOrDelete.size());
+                        return Maps.difference(rawResults, filteredResults).entriesOnlyOnLeft();
+                    })
+                    .thenApply(remainingRawResults -> getWithPostFilteringInternal(
+                            tableRef,
+                            remainingRawResults,
+                            asyncKeyValueService,
+                            asyncTransactionService,
+                            iterations + 1))
```
Comment (rhuffy) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1745]: recursion ⚠️

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 2]
```diff
@@ -1681,125 +1627,126 @@ private static boolean isSweepSentinel(Value value) {
...
+                .toImmutableMap();
+        getCounter(AtlasDbMetricNames.CellFilterMetrics.COMMIT_TS_GREATER_THAN_TRANSACTION_TS, tableRef)
+                .inc(keysToReload.size());
+
+        Map<Cell, Value> filteredResults = EntryStream.of(rawResults)
+                .removeKeys(keysToDelete::containsKey)
+                .removeKeys(keysToReload::containsKey)
+                .removeValues(value -> value.getContents().length == 0)
+                // The value has a commit timestamp less than our start timestamp, and is visible and valid.
+                .collect(MoreCollectors.entriesToCustomMap(LinkedHashMap::new));
```
Comment (rhuffy) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1708]: Extra allocation should be avoided

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 3]
```diff
@@ -1681,125 +1627,126 @@ private static boolean isSweepSentinel(Value value) {
...
+            LongLongMap commitTimestamps,
+            int iterations) {
+        Map<Cell, Long> keysToDelete = EntryStream.of(rawResults)
+                .mapValues(Value::getTimestamp)
+                .filterKeyValue((key, ts) -> commitTimestamps.getIfAbsent(ts, TransactionConstants.FAILED_COMMIT_TS)
+                        == TransactionConstants.FAILED_COMMIT_TS)
+                .toImmutableMap();
+        Map<Cell, Long> keysToReload = EntryStream.of(rawResults)
+                .removeKeys(keysToDelete::containsKey)
+                .mapValues(Value::getTimestamp)
```
Comment (rhuffy) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1692]: double iteration should be avoided

**Comment by rhuffy** [comment]
As an exercise I think this is fine, but to actually merge this, we'd probably need more than just "streams are easier to read" as a justification, especially if there's additional allocation cost in the hot read path.

**Comment by jeremyk-91** [comment]
> Could probably also separate out all the postfiltering logic into its own PostFilter class

Yep, there's actually an internal project ongoing to slice this into many parts (see #7000 or internal RFCs). I'm curious if there was something that motivated this though?

> Birthday present for @rhuffy

🎉

**Comment by rhuffy** [comment]
I was talking to Will about this refactor in general, and how some of the control flow (specifically the use of a `resultsAccumulator`) is a bit confusing for the uninitiated.

**Comment by wi11dey** [comment]
yup, motivation was just a proof-of-concept since raymond and i were discussing the code in the context of db upgrading and that it was difficult to plug in the kvs migrator, and also the code was a bit hard to follow for the uninitiated because of the various future chaining. i suggested streams and either raymond or jakub said it wouldn't be possible to use them async so this was a proof of concept using streams while keeping async behavior

the extra allocation of keysToDeleteOrReload could be removed with a map view -- we were already allocating for the result map, just renamed it to `filteredResults`. the extra iterations could probably also be removed with a little effort

I think this version is easier to reason about since there's no bouncing back and forth between getWithPostFilteringIterate/getWithPostFilteringInternal, i believe `@Output` in general should be avoided when possible so that the flow of values matches the code flow, and you need to keep less state in your head while reading through the code

anyway not suggesting we merge this right now, but the Future Spliterator approach could be something to keep in mind/reference if this code is going to be refactored anyway

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 1 above)
Comment (wi11dey) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1745]: there's a max iterations precondition already so no danger of stack overflow. the values on the stack are just references to the result maps that need to be kept alive during collecting postfiltered results in any case
